0018-logging.adoc

```diff
@@ -0,0 +1,94 @@
+
+This spike is about investigating, how the current logs could be improved and proposing an overall logging strategy that makes the logs actually useful.
+
+There are a few problems that need addressing, including but not limited to:
+
+* Log levels - guidelines on when to log and at what level
+* Useful context - we need useful context, what is useful should be defined by the team since it comes from experience debugging cnd.
+* Only useful output i.e., no dumping u8 arrays in the logs.
+* Async codebase - since we have an asynchronous code base logging may be inherently the wrong approach, we will also consider tracing.
+* Dynamically controlling log level
+* Controlling the log level of libraries vs cnd logs i.e., 
+
+The output of this spike should be proposed changes to the current logging libraries (i.e., what libraries we use to log) and also some guidelines so the whole team knows when and where to add log statements.
+
+== Research
+
+=== Log Levels ===
+
+
+This is actually the easiest thing to solve, using correct log levels is pretty standard dev practice, the internet more or less has consensus on the following log levels:
+
+Copying from the following site: https://reflectoring.io/logging-levels/
+
+* error: should only be used when the application really is in trouble. Users are being affected without having a way to work around the issue.
+* warn: should be used when something bad happened, but the application still has the chance to heal itself or the issue can wait a day or two to be fixed.
+* info: should be used to document state changes in the application or some entity within the application.
+* debug: should be used to log any information that helps us identify what went wrong on DEBUG level.
+* trace: log all information that helps us to trace the flow of the application.
+
+=== Useful Context ===
+
+Useful context is not that easy to solve due to the async nature of cnd.
+Typically one can follow the sequence of log messages to get more context than is available within the scope of the log statement.
+The solution to this is most likely to use a tracing library instead of logging, see below for more on tracing.
+
+Related to context is useful output in the log messages i.e., don't dump byte slices using "{:?}".
+
+Suggestions to remedy this include:
+* Implement custom `Debug` for any objects that we would like to print within logs.
+* Add a `Log` trait for any objects that we wish to log and call the method when logging i.e., `log::info("Got: {}", foo.log())`
+
+=== Controlling Log Levels ===
+
+Currently we can control the base log level within the config file.  More fine grained control requires code changes in `cnd/src/logging/initialize.rs`.
+If we stay with logging I would suggest that we set the cnd log level to `trace`.
+It may be useful to have a few simple scripts in tree that do typical useful grep commands on the logs (unnecessary if we go with tracing).
```

Comment:
I am not convinced about the scripts at this stage.

Reply:
Removed.